<div align="center"> 

# Investigating Oil prices and Big Tech. Companies Stock Market 

<p>Between 2001-1-1 - 2023-12-31</p>
</div>

In this project, you will apply the skills you acquired in the course to gather and wrangle real-world data with two datasets of your choice.

You will retrieve and extract the data, assess the data programmatically and visually, accross elements of data quality and structure, and implement a cleaning strategy for the data. You will then store the updated data into your selected database/data store, combine the data, and answer a research question with the datasets.

## Neccessary Imports

In [2]:
## Run this cell to successfully run this notebook
## !pip install yfinance

In [3]:
import requests
import pandas as pd
import yfinance as yf
import requests
import config

## Pipeline
1. Gather Data
2. Asses Data
3. Clean Data
4. Store Data
5. Answering Research Questions

## 1. Gather data

In this section, data using two different (data gathering methods) are extractand and then combined. Two different types of data-gathering methods are use:
- Programmatically downloading files
- Gathering data by accessing APIs

### **1.1.** Problem Statement

In this reserach, the stock prices of major tech companies (such as Apple, Microsoft, Google, Amazon, and META) and oil prices over the period spanning from January 2001 to December 2023 are investigated. By examining historical data for both sectors, the aim is to uncover insights into how changes in oil prices impact the financial performance of big tech companies, and vice versa.
The datasets would include the stock market data for:
- META
- AMazon
- Google
- Apple

### **1.2.** Gathering Data Methods and Code

In this project, two types of data gathering are use: 
- Programmatically downloading files
- Gathering data by accessing APIs

Each of the datasets have at least two variables, and have greater than 500 data samples within.

#### **Dataset 1**: Oils and Petroleum dataset 

**Type:** JSON response

**Method:** The data was gathered using the "by accessing APIs" method from https://api.eia.gov/v2/petroleum/pri/spt/data/

**Dataset variables:**

| Column Name | Description |
|---|---|
| period |  date associated with the data point |
| duoarea | identifier for a specific geographical area |
| area-name | area name |
| product | identifier for the specific petroleum product |
| product-name | name of the product |
| process | identifier for the specific process |
| process-name | The name of the process used |
| series | identifier for the specific data series |
| series-description | descriptive text name or label for the data |

### Gathering by API

In [6]:
## Gather the data using the "Gather data by accessing APIs" method from https://api.eia.gov
api_key = config.EIA_API_KEY
url = f'https://api.eia.gov/v2/petroleum/pri/spt/data/?'

params = {
    'api_key':api_key,
    'start':'2001-01-01', 
    'end':'2023-12-31'
}

response = requests.get(url, params=params)
data = response.json()
oil_data =pd.DataFrame(data['response']['data'])
oil_data.sample(5)

period duoarea      area-name   product  \
4516  2023-08-25   YCUOK             NA    EPCWTI   
152   2014-02-14   Y35NY  NEW YORK CITY  EPD2DXL0   
2520  2008-08-01     RGC             NA     EPMRU   
3511  2010-11-05     RGC             NA      EPJK   
4007  2016-07-29   Y35NY  NEW YORK CITY     EPMRU   

                           product-name process    process-name  \
4516                      WTI Crude Oil     PF4  Spot Price FOB   
152   No 2 Diesel Low Sulfur (0-15 ppm)     PF4  Spot Price FOB   
2520      Conventional Regular Gasoline     PF4  Spot Price FOB   
3511             Kerosene-Type Jet Fuel     PF4  Spot Price FOB   
4007      Conventional Regular Gasoline     PF4  Spot Price FOB   

                          series  \
4516                        RWTC   
152   EER_EPD2DXL0_PF4_Y35NY_DPG   
2520       EER_EPMRU_PF4_RGC_DPG   
3511        EER_EPJK_PF4_RGC_DPG   
4007     EER_EPMRU_PF4_Y35NY_DPG   

                                     series-description  
4516  Cushing, OK WTI Spot Price FOB (Dollars per Ba...  
152   New York Harbor Ultra-Low Sulfur No 2 Diesel S...  
2520  U.S. Gulf Coast Conventional Gasoline Regular ...  
3511  U.S. Gulf Coast Kerosene-Type Jet Fuel Spot Pr...  
4007  New York Harbor Conventional Gasoline Regular ...

In [17]:
oil_data.shape

(5000, 9)

> The gathered dataset have 5000 rows and 9 features.

In [13]:
oil_data.to_csv('data/oil_data_api.csv', index=False) ## store to cvs file

> Storing step: the dataset is saved to the local data store directory before moving to the next step.

#### **Dataset 2**: Stocks Dataset

**Type:** Pandas.DataFrame

**Method:** The data was gathered using the "Programmatically downloading files" method from Yahoo Finance using the yfinance library

**Dataset variables:**

| Column Name | Description |
|---|---|
| Date | date associated with the data point |
| Open | The opening price of the stock on that date. |
| High | The highest price reached by the stock on that date. |
| Low | The lowest price reached by the stock on that date. |
| Close | The closing price of the stock on that date. |
| Adj Close | The adjusted closing price, adjusted for dividends and splits. |
| Volume | The number of shares traded on that date. |

### Gathering by Programmatically Downloading

In [14]:
## 2nd data gathering was downloaded programitticallly from Yahoo Finance using the yfinance library
companies = ['AAPL', 'MSFT', 'GOOGL', 'AMZN', 'META']
stocks_data = yf.download(companies, start='2001-01-01', end='2023-12-31')
stocks_data.reset_index(inplace=True)
stocks_data.sample(5)

[*********************100%%**********************]  5 of 5 completed


Price        Date  Adj Close                                                \
Ticker                  AAPL       AMZN      GOOGL        META        MSFT   
4341   2018-04-06  39.862919  70.261497  50.439613  156.877625   84.311478   
374    2002-07-03   0.264959   0.712500        NaN         NaN   16.010918   
1595   2007-05-10   3.241103   3.046000  11.535059         NaN   21.949764   
4764   2019-12-10  65.218666  86.960503  67.067528  200.458069  145.104050   
3881   2016-06-08  22.718761  36.332001  37.103916  118.147209   46.733120   

Price       Close                                    ...       Open  \
Ticker       AAPL       AMZN      GOOGL        META  ...       AAPL   
4341    42.095001  70.261497  50.497501  157.199997  ...  42.742500   
374      0.313393   0.712500        NaN         NaN  ...   0.300179   
1595     3.833571   3.046000  11.548298         NaN  ...   3.808214   
4764    67.120003  86.960503  67.144501  200.869995  ...  67.150002   
3881    24.735001  36.332001  37.146500  118.389999  ...  24.754999   

Price                                                     Volume             \
Ticker       AMZN      GOOGL        META        MSFT        AAPL       AMZN   
4341    71.498497  51.154999  157.729996   91.489998   140021200  117646000   
374      0.669000        NaN         NaN   25.620001   199029600  154074000   
1595     3.122000  11.687688         NaN   30.680000  1197257600  201040000   
4764    87.370003  66.997002  201.660004  151.289993    90420400   50286000   
3881    36.320000  36.974998  117.760002   52.020000    83392400   44468000   

Price                                      
Ticker        GOOGL        META      MSFT  
4341     40134000.0  41644800.0  38026000  
374             NaN         NaN  80936600  
1595    147304548.0         NaN  55398600  
4764     17606000.0   9475700.0  16476100  
3881     32314000.0  14505600.0  21149400  

[5 rows x 31 columns]

In [18]:
stocks_data.shape

(5785, 31)

> The gathered data has 5785 data points and 31 features.

In [15]:
stocks_data.to_csv('data/stock_data.csv', index=False) ## store to cvs file

> Storing step: the dataset is saved to the local data store directory before moving to the next step.

## 2. Assess data

Assess the data according to data quality and tidiness metrics using the report below.

List **two** data quality issues and **two** tidiness issues. Assess each data issue visually **and** programmatically, then briefly describe the issue you find.  **Make sure you include justifications for the methods you use for the assessment.**


Now that we have gathered the datasets, let's assess the dataset for data quality and structural issues.

Here's a list of the data quality attributes we covered in the course for your reference:

    Completeness
    Validity
    Accuracy
    Consistency
    Uniqueness

### Quality Issue 1:

In [ ]:
## Inspecting the dataframe visually


In [ ]:
## Inspecting the dataframe programmatically


Issue and justification: *FILL IN*

### Quality Issue 2:

In [ ]:
## Inspecting the dataframe visually

In [ ]:
## Inspecting the dataframe programmatically

Issue and justification: *FILL IN*

### Tidiness Issue 1:

In [ ]:
## Inspecting the dataframe visually

In [ ]:
## Inspecting the dataframe programmatically

Issue and justification: *FILL IN*

### Tidiness Issue 2: 

In [ ]:
## Inspecting the dataframe visually

In [ ]:
## Inspecting the dataframe programmatically

Issue and justification: *FILL IN*

## 3. Clean data
It's time to address the issues found during assessment to clean and polish your data.

Clean the data to solve the 4 issues corresponding to data quality and tidiness found in the assessing step. **Make sure you include justifications for your cleaning decisions.**

After the cleaning for each issue, please use **either** the visually or programatical method to validate the cleaning was succesful.

At this stage, you are also expected to remove variables that are unnecessary for your analysis and combine your datasets. Depending on your datasets, you may choose to perform variable combination and elimination before or after the cleaning stage. Your dataset must have **at least** 4 variables after combining the data.

In [ ]:
# FILL IN - Make copies of the datasets to ensure the raw dataframes 
# are not impacted

### **Quality Issue 1: FILL IN**

In [ ]:
# FILL IN - Apply the cleaning strategy

In [ ]:
# FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Quality Issue 2: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Tidiness Issue 1: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Tidiness Issue 2: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Remove unnecessary variables and combine datasets**

Depending on the datasets, you can also peform the combination before the cleaning steps.

In [ ]:
#FILL IN - Remove unnecessary variables and combine datasets

## 4. Update your data store
Update your local database/data store with the cleaned data, following best practices for storing your cleaned data:

- Must maintain different instances / versions of data (raw and cleaned data)
- Must name the dataset files informatively
- Ensure both the raw and cleaned data is saved to your database/data store

In [ ]:
oil_data.to_csv('data/oil_data_api_cleaned.csv', index=False) ## store to cvs file

In [ ]:
stocks_data.to_csv('data/stock_data_cleaned.csv', index=False) ## store to cvs file

## 5. Answer the research question

### **5.1:** Define and answer the research question 
Going back to the problem statement in step 1, use the cleaned data to answer the question you raised. Produce **at least** two visualizations using the cleaned data and explain how they help you answer the question.

*Research question:* FILL IN from answer to Step 1

In [ ]:
#Visual 1 - FILL IN

*Answer to research question:* FILL IN

In [ ]:
#Visual 2 - FILL IN

*Answer to research question:* FILL IN

### **5.2:** Reflection
In 2-4 sentences, if you had more time to complete the project, what actions would you take? For example, which data quality and structural issues would you look into further, and what research questions would you further explore?

*Answer:* FILL IN